### PR TITLE
Add index block api

### PR DESCRIPTION
**Pull request by ywelsch**
Adds an API for putting an index block in place, which also ensures for write blocks that, once successfully returning to the user, all shards of the index are properly accounting for the block, for example that all in-flight writes to an index have been completed after adding the write block.

This API allows coordinating more complex workflows, where it is crucial that an index is no longer receiving writes after the API completes, useful for example when marking an index as read-only during an upgrade in order to reindex its documents.

Note to reviewers: I will fix the docs in a later revision once we're all good on naming etc.
Also, HLRC support will only come later.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            builder.startObject(index.getName());
+            {
+                if (hasFailures()) {
+                    builder.field("blocked", false);
+                    if (exception != null) {
+                        builder.startObject("exception");
+                        ElasticsearchException.generateFailureXContent(builder, params, exception, true);
+                        builder.endObject();
+                    } else {
+                        builder.startObject("failedShards");
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java:154]: I think this should be `failed_shards` rather than camel case to match our usual API conventions

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/TransportAddIndexBlockAction.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+import org.elasticsearch.index.Index;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Collections;
+
+/**
+ * Close index action
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/indices/readonly/TransportAddIndexBlockAction.java:46]: Documentation needs to be updated

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/TransportVerifyShardIndexBlockAction.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class TransportVerifyShardIndexBlockAction extends TransportReplicationAction<
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/indices/readonly/TransportVerifyShardIndexBlockAction.java:49]: Can you add javadocs for this class please?

File: server/src/main/java/org/elasticsearch/client/IndicesAdminClient.java [hunk 4]
```diff
@@ -284,6 +288,23 @@
      */
     void open(OpenIndexRequest request, ActionListener<OpenIndexResponse> listener);
 
+    /**
+     * Adds a block to an index
+     *
+     * @param block   The block to add
+     * @param indices The name of the indices to add the block to
+     */
+    AddIndexBlockRequestBuilder prepareAddBlock(APIBlock block, String... indices);
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/client/IndicesAdminClient.java:297]: I don't think we need to add these since we are not supporting the transport client any longer?

File: server/src/main/java/org/elasticsearch/client/support/AbstractClient.java [hunk 5]
```diff
@@ -1331,6 +1336,16 @@ public void open(final OpenIndexRequest request, final ActionListener<OpenIndexR
             execute(OpenIndexAction.INSTANCE, request, listener);
         }
 
+        @Override
+        public AddIndexBlockRequestBuilder prepareAddBlock(APIBlock block, String... indices) {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/client/support/AbstractClient.java:1340]: Same about no need to add since we aren't supporting the transport client

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 6]
```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
         return ClusterState.builder(currentState).blocks(blocks).metadata(metadata).routingTable(routingTable.build()).build();
     }
 
+    static ClusterState addIndexBlock(final Index[] indices, final Map<Index, ClusterBlock> blockedIndices,
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:313]: Can you add javadocs for this method please?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 7]
```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
...
+            if (ia != null && ia.getParentDataStream() != null && ia.getParentDataStream().getWriteIndex().getIndex().equals(index)) {
+                writeIndices.add(index.getName());
+            }
+        }
+        if (writeIndices.size() > 0) {
+            throw new IllegalArgumentException("cannot add a block to the following data stream write indices [" +
+                Strings.collectionToCommaDelimitedString(writeIndices) + "]");
+        }
+
+        clusterService.submitStateUpdateTask("add-index-block " + Arrays.toString(concreteIndices),
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:389]: Super minor, but perhaps
```suggestion
        clusterService.submitStateUpdateTask("add-index-block-[" + request.getBlock() + "]-" Arrays.toString(concreteIndices),
```

So we can see the intended block in the pending tasks list

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 8]
```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
...
+                public void clusterStateProcessed(final String source, final ClusterState oldState, final ClusterState newState) {
+                    if (oldState == newState) {
+                        assert blockedIndices.isEmpty() : "List of blocked indices is not empty but cluster state wasn't changed";
+                        listener.onResponse(new AddIndexBlockResponse(true, false, Collections.emptyList()));
+                    } else {
+                        assert blockedIndices.isEmpty() == false : "List of blocked indices is empty but cluster state was changed";
+                        threadPool.executor(ThreadPool.Names.MANAGEMENT)
+                            .execute(new WaitForBlocksApplied(blockedIndices, request,
+                                    ActionListener.wrap(verifyResults ->
+                                            clusterService.submitStateUpdateTask("finalize-index-block",
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:409]: Can you add the block type and `blockedIndices` list to this state update task as well?

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequestBuilder.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+package org.elasticsearch.action.admin.indices.readonly;
+
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.support.master.AcknowledgedRequestBuilder;
+import org.elasticsearch.client.ElasticsearchClient;
+import org.elasticsearch.cluster.metadata.IndexMetadata.APIBlock;
+
+/**
+ * Builder for close index request
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequestBuilder.java:28]: Docstrings in this file need to be updated :)

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 10]
```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
...
+                    }
+                }
+            }
+            if (indexBlock == null) {
+                // Create a new UUID-based block
+                indexBlock = createUUIDBasedBlock(block.block);
+            }
+            assert Strings.hasLength(indexBlock.uuid()) : "Block should have a UUID";
+            blocks.addIndexBlock(index.getName(), indexBlock);
+            blockedIndices.put(index, indexBlock);
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:352]: I'm really uncomfortable mutating the `blockedIndices` map that was passed in as a parameter. I had to read this over and over to see what was happening. Can we refactor this method so that it becomes part of the return value (maybe using a `Tuple` if we don't want to add an object) so that it's easier to reason about?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 11]
```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
...
+                    .put(indexMetadata.getSettings()).put(block.settingName(), true).build();
+
+                metadata.put(IndexMetadata.builder(indexMetadata)
+                    .settings(updatedSettings)
+                    .settingsVersion(indexMetadata.getSettingsVersion() + 1));
+            }
+        }
+
+        logger.info(() -> new ParameterizedMessage("adding block {} to indices {}", block.name,
+            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.joining(","))));
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:366]: Why use a lambda and parameterized message here? Usually it's used when also passing an exception we want to preserve as a `cause`. Am I missing something?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 12]
```diff
@@ -304,6 +310,149 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
...
+        return ClusterState.builder(currentState).blocks(blocks).metadata(metadata).routingTable(routingTable.build()).build();
+    }
+
+    public void addIndexBlock(AddIndexBlockClusterStateUpdateRequest request,
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:370]: Could you add javadocs please?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 13]
```diff
@@ -429,6 +578,108 @@ public void onFailure(Exception e) {
         }
     }
 
+    class WaitForBlocksApplied extends ActionRunnable<Map<Index, AddBlockResult>> {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:581]: Javadocs please :)

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 14]
```diff
@@ -618,6 +869,60 @@ ClusterState openIndices(final Index[] indices, final ClusterState currentState)
         return ClusterState.builder(updatedState).routingTable(routingTable.build()).build();
     }
 
+    static Tuple<ClusterState, Collection<AddBlockResult>> finalizeBlock(final ClusterState currentState,
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:872]: Javadocs please :) they really do help when code spelunking in the future (at least I find them incredibly useful)

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for adding this Yannick, I think this will be a neat API and it'll be nice to be able to use this from ILM. I left some comments for you. I have two additional thoughts:

- I think we should change `/{index}/_blocks/{block}` to `/{index}/_block/{block}`. By and large the bulk of our APIs use singular verbs/nouns, and it reads better mentally (`/foo/_blocks/read` can be thought of as "foo blocks reads", but `/foo/_blocks/read_only` is awkward thought of as "foo blocks read onlys", it's nicer to think of `/foo/_block/read_only` as "add the read_only *block* to foo"). It's also useful because we only support a single block, not a comma-separated or wildcard of block**s** to add to the index.
- How do these blocks work for a searchable snapshot index? Should we add a test for that? (the write block is pretty obviously not needed, but a read block may still be desired)

**Comment by ywelsch** [comment]
@dakrone I've adapted `/{index}/_blocks/{block}` to `/{index}/_block/{block}` according to your suggestion. Regarding the interoperability with searchable snapshots: The API and verify action here is completely agnostic to the underlying engine being used, so will just work with frozen / searchable snapshots / whatever other flavor we have there. I will comment in-line on the other stuff.

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequest.java [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+        this.indices = indices;
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (CollectionUtils.isEmpty(indices)) {
+            validationException = addValidationError("index is missing", validationException);
+        }
+        return validationException;
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequest.java:66]: I've made it so that it never can be null (as also requested below)

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequest.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+
+    public AddIndexBlockRequest(StreamInput in) throws IOException {
+        super(in);
+        indices = in.readStringArray();
+        indicesOptions = IndicesOptions.readIndicesOptions(in);
+        block = APIBlock.readFrom(in);
+    }
+
+    public AddIndexBlockRequest() {
+    }
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequest.java]: I've removed it

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequestBuilder.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+ */
+public class AddIndexBlockRequestBuilder
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockRequestBuilder.java:30]: I would prefer to keep the builder, in particular as this is being backported to 7.x, which still has the transport client. I also prefer to keep things consistent in the code base.

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            builder.startObject(index.getName());
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java:145]: Looking at other similar APIs here (close index, or index stats), they all dynamically key by index name. For consistency reasons, I prefer to go with what we currently have today. I understand that we want to eventually move to a different place, but would prefer doing so consistently across these APIs.

File: server/src/main/java/org/elasticsearch/client/IndicesAdminClient.java (hunk 4 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/client/IndicesAdminClient.java:297]: see comment above about transport client

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 10 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:361]: Good idea, changed  this to a tuple

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java (hunk 11 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java]: No reason, code was copy-pasted from elsewhere

**Comment by ywelsch** [comment]
@jrodewig can you help out with the docs here? (Feel free to directly push changes to this PR)

File: docs/reference/index-modules/blocks.asciidoc [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+==== {api-path-parms-title}
+
+include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=index]
++
+To add blocks to all indices, use `_all` or `*`. To disallow the adding
+of blocks to indices with `_all` or wildcard expressions,
+change the `action.destructive_requires_name` cluster setting to `true`.
+You can update this setting in the `elasticsearch.yml` file
+or using the <<cluster-update-settings,cluster update settings>> API.
+
```
Comment (jrodewig) [inline docs/reference/index-modules/blocks.asciidoc:97]: We should explicitly document the `<block>` path parameter.

```suggestion
`<block>`::
(Required, string)
Block type to add to the index.
+
.Valid values for `<block>`
[%collapsible%open]
====
`metadata`::
Disable metadata changes, such as closing the index.

`read`::
Disable read operations.

`read_only`::
Disable write operations and metadata changes.

`read_only_allow_delete`::
Disable write operations and metadata changes.
Document deletion is disabled.
However, index deletion is still allowed.

`write`::
Disable write operations. However, metadata changes are still allowed.
====
```

File: docs/reference/index-modules/blocks.asciidoc [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+adding the write block.
+
+[source,console]
+--------------------------------------------------
+PUT /twitter/_block/write
+--------------------------------------------------
+// TEST[setup:twitter]
+
+The blocks available are `read_only`, `read_only_allow_delete`, `read`, `write`,
+and `metadata`, and their meaning is described above in the settings section.
```
Comment (jrodewig) [inline docs/reference/index-modules/blocks.asciidoc:64]: I'd move this to the `<block>` path parameter definition below.
```suggestion
```

File: docs/reference/index-modules/blocks.asciidoc [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+    Set to `true` to disable index metadata reads and writes.
+
+[discrete]
+[[add-index-block]]
+=== Add index block API
+
+Adds an index block, and ensures that, once successfully returning to the
+user, all shards of the index are properly accounting for the block, for
+example that all in-flight writes to an index have been completed after
+adding the write block.
```
Comment (jrodewig) [inline docs/reference/index-modules/blocks.asciidoc:55]: The second part of this sentence (`, and ensures that...`) confused me. Is it saying `If successful, the API returns a confirmation`?  Either way, I'm not sure it's needed. 

If you decide to keep it, I would reword and split into a separate sentence.
```suggestion
Adds an index block to an index.
```

File: docs/reference/index-modules/blocks.asciidoc [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+
+`index.blocks.read`::
+
+    Set to `true` to disable read operations against the index.
+
+`index.blocks.write`::
+
+    Set to `true` to disable data write operations against the index. Unlike `read_only`,
+    this setting does not affect metadata. For instance, you can close an index with a `write`
+    block, but not an index with a `read_only` block.
```
Comment (jrodewig) [inline docs/reference/index-modules/blocks.asciidoc:42]: Tweak for clarity.
```suggestion
    block, but you cannot close an index with a `read_only` block.
```

File: docs/reference/index-modules/blocks.asciidoc [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+
+[discrete]
+[[add-index-block-api-wait-for-active-shards]]
+===== Wait For active shards
+
+include::{es-repo-dir}/indices/open-close.asciidoc[tag=wait-for-active-shards]
+
+[discrete]
+[[add-index-block-api-path-params]]
+==== {api-path-parms-title}
```
Comment (jrodewig) [inline docs/reference/index-modules/blocks.asciidoc:88]: This ensures the `<block>` path parameter definition renders correctly.
```suggestion
[discrete]
[role="child_attributes"]
[[add-index-block-api-path-params]]
==== {api-path-parms-title}
```

**Review by jrodewig — COMMENTED** [review]
@ywelsch I made some changes to fix the docs CI test. However, I left some other feedback that should be addressed.

**Review by jrodewig — APPROVED** [review]
:doh: Forgot you gave me permission to push the changes. I went ahead and added my suggestions. 😄

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java [hunk 24]
```diff
@@ -304,6 +310,176 @@ static ClusterState addIndexClosedBlocks(final Index[] indices, final Map<Index,
...
+                    .put(indexMetadata.getSettings()).put(block.settingName(), true).build();
+
+                metadata.put(IndexMetadata.builder(indexMetadata)
+                    .settings(updatedSettings)
+                    .settingsVersion(indexMetadata.getSettingsVersion() + 1));
+            }
+        }
+
+        logger.info("adding block {} to indices {}", block.name,
+            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.joining(",")));
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:375]: super minor, but since we usually wrap things in `[]`:

```suggestion
            blockedIndices.keySet().stream().map(Object::toString).collect(Collectors.toList()));
```

File: docs/reference/index-modules/blocks.asciidoc [hunk 25]
```diff
@@ -0,0 +1,146 @@
+[[index-modules-blocks]]
+== Index blocks
+
+Index blocks limit the kind of operations that are available on a certain
+index. The blocks come in different flavours, allowing to block write,
+read, or metadata operations. The blocks can be set / removed using dynamic
+index settings, or can be added using a dedicated API, which also ensures
+that, once successfully returning to the user, all shards of the index are
+properly accounting for the block, for example that all in-flight writes to
+an index have been completed after adding the write block.
```
Comment (dakrone) [inline docs/reference/index-modules/blocks.asciidoc:10]: Should we call out that this only applies to write operations, not read operations?

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks for the update Yannick, I made some more comments.

As an aside, it doesn't appear to me (though I am certainly not an expert in shard operation permits!) that this allows adding a read block and waiting for all query operations to finish before returning. Is that correct? Out of curiosity, how hard would that be to add? (It wouldn't have to be in this PR) It would be very useful for ILM operations if we had that ability

File: server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/admin/indices/readonly/AddIndexBlockResponse.java:145]: I've adapted the PR to your wishes now, but I don't agree with the general principle. What the team is suggesting is making our lives easier (less stuff to transition), but at the same time is leaving the user with an inconsistent set of APIs (every API having a slightly different look and feel). I think it's the wrong tradeoff to do, especially with the versioned HTTP API looming around the corner (where the need to strive for consistency within the same API version should be an even stronger guiding principle).

**Comment by ywelsch** [comment]
> As an aside, it doesn't appear to me (though I am certainly not an expert in shard operation permits!) that this allows adding a read block and waiting for all query operations to finish before returning. Is that correct?

You're absolutely right here. The PR was overpromising in this regard (I initially started with a dedicated API to just allow adding a write block, which eventually got turned into this more general one here). I've adapted the docs to point out that we only guarantee this for write operations.

> Out of curiosity, how hard would that be to add? (It wouldn't have to be in this PR) It would be very useful for ILM operations if we had that ability

We could do something similar for reads as we do for writes, but there are a bigger number of cases / scenarios to consider and to handle accordingly. For example, how should his affect scrolls that have been opened before the block has been put in place (AFAICS we don't directly close the scroll context today). Similarly how should this handle long-running searches, or a fetch phase (where the query phase successfully executed)...

**Review by dakrone — APPROVED** [review]
LGTM, thanks for iterating on it Yannick!

> For example, how should his affect scrolls that have been opened before the block has been put in place (AFAICS we don't directly close the scroll context today). Similarly how should this handle long-running searches, or a fetch phase (where the query phase successfully executed)...

This is the exact scenario I was thinking about, in some places in ILM (like after a shrink) we do a swap-and-delete from the alias, but we have some unavoidable issues where the query may be in the middle of execution during the delete, we unfortunately don't have a way to wait until they are all done. Maybe in the future though?

**Comment by ywelsch** [comment]
> in some places in ILM (like after a shrink) we do a swap-and-delete from the alias, but we have some unavoidable issues where the query may be in the middle of execution during the delete, we unfortunately don't have a way to wait until they are all done.

I'm wondering if that's a situation which should rather be covered by a more resilient _search API, where it would retry searching on the new shards (after alias been swapped), similar to what's discussed in https://github.com/elastic/elasticsearch/issues/56236

**Comment by ywelsch** [comment]
Thanks @dakrone and @jrodewig
